cxx4opencl/address_spaces.txt

```diff
@@ -209,46 +292,61 @@ __kernel void foo() {
+copy or move assignment, destructor) has the default address space for an implicit
 object pointer and reference parameters (see also
-<<addrspace-member-function-qualifiers, _Member function qualifier_>>).
+<<addrspace-member-function-qualifiers, _Member function qualifier_>>). This default
+address space is generic if it is supported or private otherwise.
 
 [source,cpp]
-----------
+------------
 class C {
   // Has the following implicitly defined member functions.
 
-  // void C(); /* implicit 'this' parameter is a pointer to */
-               /* object in generic address space. */
+  // C(); /* implicit 'this' parameter is a pointer to */
+          /* object in generic address space if supported */
+          /* in private address space otherwise. */
 
-  // void C(const C & par); /* 'this'/'par' is a pointer/reference to */
-                            /*  object in generic address space. */
+  // C(const C & par); /* 'this'/'par' is a pointer/reference to */
+                       /*  object in generic address space */
+                       /*  if supported */
+                       /* in private address space otherwise. */
 
-  // void C(C && par); /* 'this'/'par' is a pointer/r-val reference to */
-                       /* object in generic address space. */
+  // C(C && par); /* 'this'/'par' is a pointer/r-val reference to */
+                  /* object in generic address space if supported */
+                  /* in private address space otherwise. */
 
-  // operator= C &(const C & par); /* 'this'/'par'/return value is */
+  // C & operator=(const C & par); /* 'this'/'par'/return value is */
                                    /* a pointer/reference/reference to */
-                                   /* object in generic address space. */
+                                   /* object in generic address space */
+                                   /* if supported */ 
+                                   /* in private address space otherwise. */
 
-  // operator= C &(C && par)'; /* 'this'/'par'/return value is */
+  // C & operator=(C && par)'; /* 'this'/'par'/return value is */
                                /* a pointer/r-val reference/reference to */
                                /* object in generic address space. */
+                               /* if supported */ 
+                               /* in private address space otherwise. */
 };
-----------
+------------
+
+`TODO: As an optimisation we can do an upgrade in this version and always infer
+the concrete address space! See "Builtin operators" for details. Needs
+confirmation that code size doesn't increase significantly.`
 
 ==== Builtin operators
 
 All builtin operators are available in the specific named address spaces, thus
-no conversion to generic address space is performed.
+no address space conversions (i.e. to generic address space) are performed.
+
+`TODO: Why are assignment operators different? See "Implicit special members".`
```

Comment:
Is this TODO meant to stay and be published?

Reply:
This can be resolved now because in C++17 [over.built] p1 it states:

> Because built-in operators take only operands with
non-class type, ...

From which there seems to be no ambiguity between member assignments and built-in assignment operators as they are classified in different categories.